### PR TITLE
Set nanny environment variables in config

### PR DESCRIPTION
**Pull request by mrocklin**
With the nanny we have the ability to set environment variables in dask workers.
This makes this configurable with the dask.config system.

Somewhat more controversially, this also sets a few common defaults like
MKL_NUM_THREADS, OMP_NUM_THREADS, and MALLOC_TRIM_THRESHOLD

These *may* have surprising effects, however I suspect that setting
these defaults will help far more people than it would harm.

cc @jacobtomlinson @crusaderky @fjetter and @jcrist who I think are the most likely to be concerned by the default

### CODE REVIEW

**Review by fjetter — APPROVED** [review]
> These may have surprising effects, however I suspect that setting
these defaults will help far more people than it would harm.

I'm on board. You're probably right.

---

What's your reasoning for setting the num threads arguments? That might upset some folks and I don't see as big of a win for them as for the malloc setting.

File: distributed/distributed.yaml
Comment (crusaderky) [inline distributed/distributed.yaml:146]:
```suggestion
      # Linux only; inconsequential on Windows, MacOS, or when using alternative memory managers.
      # See https://distributed.dask.org/en/latest/worker.html#memory-not-released-back-to-the-os
      MALLOC_TRIM_THRESHOLD_: 65536
```

File: distributed/nanny.py
Comment (crusaderky) [inline distributed/nanny.py:165]:
```suggestion
        self.env = dask.config.get("distributed.nanny.environ")
        self.env.update(env or {})
```
Also I could not find anything that passes the env parameter to the Nanny constructor. Maybe we could deprecate it? Unless some dask users actually create Nanny objects by hand?

File: distributed/tests/test_nanny.py
Comment (crusaderky) [inline distributed/tests/test_nanny.py:380]:
```suggestion
        results = await c.run(os.getenv, "ABC")
```

File: distributed/tests/test_nanny.py
Comment (crusaderky) [inline distributed/tests/test_nanny.py:381]:
```suggestion
    config={"distributed.nanny.environ": {"A": 1, "B": 2}},
)
async def test_environment_variable_config(c, s):
    async with Nanny(s.address, env={"B": 4, "C": 3}) as n:
        results = await c.run(os.getenv, "A")
        assert results[n.worker_address] == "1"
        results = await c.run(os.getenv, "B")
        assert results[n.worker_address] == "4"
        results = await c.run(os.getenv, "C")
        assert results[n.worker_address] == "3"
```

File: distributed/distributed.yaml
Comment (mrocklin) [reply distributed/distributed.yaml:146]: Historically we've avoided putting lots of comments in the main config file, relying instead on the description in the schema file (which gets put into docs).  I'm also OK being silent about this because it doesn't have any negative effect.

Comment (jrbourbeau) [reply distributed/distributed.yaml:146]: What's the reasoning behind this value for `MALLOC_TRIM_THRESHOLD_`?

**Review by jrbourbeau — COMMENTED** [review]
I'd like to hold off on merging this until after the release this Friday. That will give us a bit more time to hear from folks using the `main` branch of `distributed`

**Comment by mrocklin** [comment]
> What's the reasoning behind this value for MALLOC_TRIM_THRESHOLD_?

@crusaderky suggested it, and I trust @crusaderky  :)

File: distributed/nanny.py
Comment (mrocklin) [reply distributed/nanny.py]: I'm inclined to leave the deprecation question to other work if that's ok.  It seems harmless to leave here for now and there are other things that seem higher priority.

**Comment by mrocklin** [comment]
> What's your reasoning for setting the num threads arguments? That might upset some folks and I don't see as big of a win for them as for the malloc setting.

Anyone using blas/lapack today essentially needs to set these in order to avoid massively oversubscribing their machines.  OpenBLAS in particular will often just shut down the program if you try to use more than twice the number of threads on the system.

File: distributed/nanny.py
Comment (jacobtomlinson) [inline distributed/nanny.py:165]: Should we also pull in the local environment to ensure that takes precident?

```suggestion
        self.env = {**dask.config.get("distributed.nanny.environ"), **env, **os.environ}
```

**Review by jacobtomlinson — COMMENTED** [review]
This generally sounds good to me.

My only concern is that this feels a little magic and I think it could be easy for a user to not understand where an environment variable is coming from. It would be easy to set something in your YAML and forget about it.

For example if I set:

```yaml
distributed:
  nanny:
    environ:
      FOO: "bar"
```

and then run

```
$FOO=baz dask-worker tcp://localhost:8786
```

The environment variable on the worker will be `bar`. This doesn't feel intuitive.

File: distributed/nanny.py
Comment (mrocklin) [reply distributed/nanny.py:165]: I think that precedence here becomes interesting.  I would certainly want `**env` to take precedence.  Even between config and os.environ it's an open question.  If someone sets an environment variable on the command line then sure, but what about if it's set in bashrc?  I might think differently then.

Comment (mrocklin) [reply distributed/nanny.py]: However, I've made an edit with something similar to your suggestion.  I'm ambivalent on this point.

**Comment by crusaderky** [comment]
> > What's the reasoning behind this value for MALLOC_TRIM_THRESHOLD_?
> 
> @crusaderky suggested it, and I trust @crusaderky :)

0 or 65536 is pretty much the same for CPython and CPython extenions like numpy, because PyMalloc invokes the underlying malloc() in large-ish chunks.
However, C/C++ libraries that invoke malloc directly - which are the whole reason why the glibc implements its own user-space memory manager - would risk getting severely harmed by setting the trim to 0.

**Comment by mrocklin** [comment]
@jrbourbeau are we ok to merge this in?

**Review by jrbourbeau — APPROVED** [review]
Thanks @mrocklin. It looks like there's no objections to setting these default environment variables. Will merge once CI finishes

@jacobtomlinson, to your point about environment variable priority, Matt included an update where environment variables in `os.environ` will take priority over those in the `distributed.nanny.environ` configuration setting. I _think_ this addresses the concerns you raised, but if not I'm happy to follow up in a separate PR

**Pull request merged**
